File: source/android/src/eu/alebianco/air/extensions/utils/FREUtils.java
/**
 * Project: ANE-Google-Analytics
 *
 * Author:  Alessandro Bianco
 * Website: http://alessandrobianco.eu
 * Twitter: @alebianco
 * Created: 21/12/12 14.08
 *
 * Copyright © 2013 Alessandro Bianco
 */
package eu.alebianco.air.extensions.utils;

import android.util.Log;
import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;

import java.io.PrintWriter;
import java.io.StringWriter;

public class FREUtils {

    public static final String TAG = "ANE";

    public static void logEvent(FREContext context, LogLevel level, String format, Object... args) {
        String message = String.format(format, args);
        Log.println(level.getPriority(), TAG, message);
        context.dispatchStatusEventAsync(message, level.getName());
    }
    
    public static String stripPackageFromClassName(String className) {
    	return className.substring(className.lastIndexOf(".") + 1);
    }

	public static FREObject createRuntimeException(String className, int id, String message, Object... data) {
		FREObject error = null;
        FREObject[] args = new FREObject[2] ;
        try {
        	args[0] = FREObject.newObject(String.format(message, data));
            args[1] = FREObject.newObject(id);
            error = FREObject.newObject(className, args);
		} catch (Exception e) {
			Log.e(TAG, "Inception Error: Unable to create the runtime exception to notify the application about the previous error.");
		}
        return error;
	}

    public static String getStackTrace(Throwable exception) {
        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        exception.printStackTrace(pw);
        return sw.toString();

    }
}

File: source/ios/NativeGATracker/FREUtils.m
/**
 * Project: ANE-Google-Analytics
 *
 * Author:  Alessandro Bianco
 * Website: http://alessandrobianco.eu
 * Twitter: @alebianco
 * Created: 26/12/12 16.06
 *
 * Copyright © 2013 Alessandro Bianco
 */

#import "FREUtils.h"
#import "FREConversionUtil.h"

@implementation FREUtils

void logEvent(FREContext ctx, enum LogLevel lvl, NSString *format, ...) {
    va_list ap = NULL;
    va_start(ap, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:ap];
    va_end(ap);

    NSString *level = NULL;
    switch (lvl) {
        case kInfo:
            level = @"INFO";
            break;
        case kDebug:
            level = @"DEBUG";
            break;
        case kWarn:
            level = @"WARN";
            break;
        case kError:
            level = @"ERROR";
            break;
        case kFatal:
            level = @"FATAL";
            break;

        default:
            level = @"INFO";
            break;
    }

    NSLog(@"%@: %@", level, message);
    DISPATCH_EVENT(ctx, (uint8_t *) [[@"INTERNAL_" stringByAppendingString:level] UTF8String], (uint8_t *) [message UTF8String]);
}
FREObject createRuntimeException(NSString *type, NSInteger id, NSString *format, ...) {
    FREObject object;

    va_list ap = NULL;
    va_start(ap, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:ap];
    va_end(ap);

    uint32_t argc = 2;

    FREObject argv[argc];
    argv[0] = [FREConversionUtil fromString:message];
    argv[1] = [FREConversionUtil fromInt:id];

    FREResult result = FRENewObject((uint8_t *) [type UTF8String], argc, argv, &object, NULL);

    if (result == FRE_OK) {
        return object;
    } else {
        NSLog(@"Inception Error: Unable to create the runtime exception to notify the application about the previous error.");
    }

    return NULL;
}

@end
